src/vec_int16_ppc.h

```diff
@@ -0,0 +1,392 @@
+ * This header serves to fill in functional gaps for older
+ * (POWER7, POWER8) processors and provides an in-line assembler
+ * implementation for older compilers that do not
+ * provide the build-ins.
+ *
+ * This header covers operations that are either:
+ *
+ * - Implemented in hardware instructions for later
+ * processors and useful to programmers, on slightly older processors,
+ * even if the equivalent function requires more instructions.
+ * Examples include Count Leading Zeros, Population Count and Byte
+ * Reverse.
+ * - Defined in the OpenPOWER ABI but <I>not</I> yet defined in
+ * <altivec.n> provided by available compilers in common use.
+ * Examples include Count Leading Zeros, Population Count and Byte
+ * Reverse.
+ * - Commonly used operations, not covered by the ABI or
+ * <altivec.h>, and require multiple instructions or
+ * are not obvious.
+ * Examples include the shift immediate operations.
+ *
+ */
+
+
+/** \brief Count leading zeros for a vector unsigned short (halfword)
+ *  elements.
+ *
+ *  Count the number of leading '0' bits (0-16) within each halfword
+ *  element of a 128-bit vector.
+ *
+ *  For POWER8 (PowerISA 2.07B) or later use the Vector Count Leading
+ *  Zeros Halfword instruction <B>vclzh</B>. Otherwise use sequence of pre
+ *  2.07 VMX instructions.
+ *  SIMDized count leading zeros inspired by:
+ *
+ *  Warren, Henry S. Jr and <I>Hacker's Delight</I>, 2nd Edition,
+ *  Addison Wesley, 2013. Chapter 5 Counting Bits, Figure 5-12.
+ *
+ *  @param vra 128-bit vector treated as 8 x 16-bit integer (halfword)
+ *  elements.
+ *  @return 128-bit vector with the Leading Zeros count for each
+ *  halfword element.
+ */
+static inline vui16_t
+vec_clzh (vui16_t vra)
+{
+  vui16_t r;
+#ifdef _ARCH_PWR8
+#ifndef vec_vclzh
+  __asm__(
+      "vclzh %0,%1;"
+      : "=v" (r)
+      : "v" (vra)
+      : );
+#else
+  r = vec_vclzh (vra);
+#endif
+#else
+//#warning Implememention pre power8
+  vui16_t n, nt, y, x, s, m;
+  vui16_t z= { 0,0,0,0, 0,0,0,0};
+  vui16_t one = { 1,1,1,1, 1,1,1,1};
+
+  /* n = 16 s = 8 */
+  s = vec_splat_u16(8);
+  n = vec_add (s, s);
+  x = vra;
+
+  /* y=x>>8 if (y!=0) (n=n-8 x=y)  */
+  y = vec_sr(x, s);
+  nt = vec_sub(n,s);
+  m = (vui16_t)vec_cmpgt(y, z);
+  s = vec_sr(s,one);
+  x = vec_sel (x , y, m);
+  n = vec_sel (n , nt, m);
+
+  /* y=x>>4 if (y!=0) (n=n-4 x=y)  */
+  y = vec_sr(x, s);
+  nt = vec_sub(n,s);
+  m = (vui16_t)vec_cmpgt(y, z);
+  s = vec_sr(s,one);
+  x = vec_sel (x , y, m);
+  n = vec_sel (n , nt, m);
+
+  /* y=x>>2 if (y!=0) (n=n-2 x=y)  */
+  y = vec_sr(x, s);
+  nt = vec_sub(n,s);
+  m = (vui16_t)vec_cmpgt(y, z);
+  s = vec_sr(s,one);
+  x = vec_sel (x , y, m);
+  n = vec_sel (n , nt, m);
+
+  /* y=x>>1 if (y!=0) return (n=n-2)   */
+  y = vec_sr(x, s);
+  nt = vec_sub(n,s);
+  nt = vec_sub(nt,s);
+  m = (vui16_t)vec_cmpgt(y, z);
+  n = vec_sel (n , nt, m);
+
+  /* else return (x-n)  */
+  nt = vec_sub (n, x);
+  r = vec_sel (nt , n, m);
+#endif
+
+  return (r);
```

Comment:
nit: don't need the parens here.  Also in many other places.  not sure how much to care.